File: src/main/scala/com/github/tetrisanalyzer/settings/DefaultGameSettings.scala
package com.github.tetrisanalyzer.settings

import com.github.tetrisanalyzer.piece.{Piece}
import com.github.tetrisanalyzer.move.rotation.AnticlockwiseRotation
import com.github.tetrisanalyzer.move.Move
import java.awt.Color

object DefaultGameSettings {
  private val Colors = Array(
    new Color(200, 200, 230), // 0 = Empty
    new Color(255,0,0),       // 1 = I, Red
    new Color(255, 151, 0),   // 2 = Z, Orange
    new Color(0, 184, 151),   // 3 = S, Cyan
    new Color(222, 184, 0),   // 4 = J, Yellow
    new Color(222, 0, 222),   // 5 = L, Purple
    new Color(71, 184, 0),    // 6 = T, Green
    new Color(0, 71, 222),    // 7 = O, Blue
    new Color(100, 100, 130), // 8 = Any Piece
    new Color(200, 200, 200)  // 9 = Wall
  )

  private val LineColors = Array.tabulate(Colors.size) {
      ((i) => (new Color(
        Colors(i).getRed / 2,
        Colors(i).getGreen / 2,
        Colors(i).getBlue / 2)))
  }

  Colors(0) = new Color(255, 255, 255);
}

/**
 * The default settings used by the Tetris engine.
 */
class DefaultGameSettings extends GameSettings {
  def isSlidingEnabled = false
  val rotationDirection = new AnticlockwiseRotation
  def pieceStartMove(boardWidth: Int, piece: Piece) = new Move(0, if (boardWidth <= 4) 0 else (boardWidth - 2) / 2, 0)

  def squareColor(element: Int) = DefaultGameSettings.Colors(element)
  def lineColor(element: Int) = DefaultGameSettings.LineColors(element)
}

File: langs/java/src/main/java/com/github/tetrisanalyzer/settings/DefaultGameSettings.java
package com.github.tetrisanalyzer.settings;

import com.github.tetrisanalyzer.move.rotation.AnticlockwiseRotation;
import com.github.tetrisanalyzer.move.rotation.RotationDirection;
import com.github.tetrisanalyzer.piece.Piece;

import static com.github.tetrisanalyzer.piece.Piece.*;

/**
 * The default settings used by the Tetris engine.
 */
public class DefaultGameSettings implements GameSettings {

    private final int boardWidth;
    private final int boardHeight;
    private final int pieceStartX;

    public DefaultGameSettings() {
        this(10, 20, 4);
    }

    public DefaultGameSettings(int boardWidth, int boardHeight, int pieceStartX) {
        this.boardWidth = boardWidth;
        this.boardHeight = boardHeight;
        this.pieceStartX = pieceStartX;
    }

    public boolean isSlidingEnabled() {
        return false;
    }

    public RotationDirection rotationDirection() {
        return new AnticlockwiseRotation();
    }

    @Override
    public int pieceStartX() {
        return pieceStartX;
    }

    @Override
    public int boardWidth() {
        return boardWidth;
    }

    @Override
    public int boardHeight() {
        return boardHeight;
    }

    @Override
    public int[] startPieceDx() {
        return new int[] { 0, 0, 0, 0, 0, 0, 0, 0, 0 };
    }

    @Override
    public int[] startPieceDy() {
        return new int[] { 0, 0, 0, 0, 0, 0, 0, 0, 0 };
    }

    @Override
    public int[][] pieceDx() {
        int[][] result = new int[Piece.NUMBER_OF_PIECE_TYPES + 2][];
        result[I] = new int[] { 0, 0 };
        result[Z] = new int[] { 0, 0 };
        result[S] = new int[] { 0, 0 };
        result[J] = new int[] { 0, 0, 0, 0 };
        result[L] = new int[] { 0, 0, 0, 0 };
        result[T] = new int[] { 0, 0, 0, 0 };
        result[O] = new int[] { 0 };
        return result;
    }

    @Override
    public int[][] pieceDy() {
        int[][] result = new int[Piece.NUMBER_OF_PIECE_TYPES + 2][];
        result[I] = new int[] { 0, 0 };
        result[Z] = new int[] { 0, 0 };
        result[S] = new int[] { 0, 0 };
        result[J] = new int[] { 0, 0, 0, 0 };
        result[L] = new int[] { 0, 0, 0, 0 };
        result[T] = new int[] { 0, 0, 0, 0 };
        result[O] = new int[] { 0 };
        return result;
    }
}
